Adds support for Multiple products

nullbear opened the pull request:
Adds support for multiple recipe results.
Makes it so that recipes can have multiple biproducts, instead of just
one. The results list works just like the reagents list.

ex. list("reagentid" = 5, "otherreagent" = 3) where 5 and 3 are the
amounts produced.

ALSO ADDED SUPPORT FOR COLD REACTIONS. IE. cooling a mix to react, instead of heating it.

recipe on_reaction 'created volume' is the multiplier for the recipe,
not the sum of the volume of the biproducts. Equations have been
adjusted to this change.

A couple of reactions are made slightly more efficient.

Removes a couple of trailing returns and unneccessary variable
declarations.

This is a refactor.

nullbear committed:
Adds support for multiple recipe results.

Makes it so that recipes can have multiple biproducts, instead of just
one. The results list works just like the reagents list.

ex. list("reagentid" = 5, "otherreagent" = 3) where 5 and 3 are the
amounts produced.

recipe on_reaction 'created volume' is the multiplier for the recipe,
not the sum of the volume of the biproducts. Equations have been
adjusted to this change.

A couple of reactions are made slightly more efficient.

Removes a couple of trailing returns and unneccessary variable
declarations.

Scales cfl3 fireball a miniscule amount with created volume. the range
is increased by 1 for every 100 units mixed.

nullbear committed:
updates pyrotechnics.dm

PKPenguin321 commented:
I thought this would be about ~~table~~crafting but this is really good too

nullbear commented:
Tablecrafting would b hard though cause it uses a reference of the product and tries to put stuff in it.

phil235 commented on code/modules/reagents/chemistry/holder.dm:
```diff
@@ -382,11 +382,10 @@ var/const/INJECT = 5 //injection
-					var/created_volume = C.result_amount*multiplier
-					if(C.result)
-						feedback_add_details("chemical_reaction","[C.result]|[C.result_amount*multiplier]")
+					for(var/result in C.results)
```
no a great var name, easy to get confused.

phil235 commented on code/modules/reagents/chemistry/recipes/others.dm:
```diff
@@ -2,211 +2,183 @@
 	name = "Space Lube"
 	id = "lube"
-	result = "lube"
+	results = list("lube" = 3) // fuck this spillover into nullspace.
```
These comments are not needed, they don't say anything about the code.

nullbear committed:
Removes unneccessary comments and unrelated tweaks.

nullbear committed:
Changes variable name for phil

nullbear committed:
P is for Product. R is for Reagent.

nullbear committed:
Adds support for inverse required temp reactions.

Ie. Cooling a mixture to cause a reaction, instead of heating it. Just
set 'is_cold_recipe' to 1 in the reaction.

nullbear commented:
Oh dear i added an infinite loop somewhere.

nullbear committed:
Fixes chapel exploding infinite loop.

Fixes weird bug. Adds unused colorless 'powderpile' icon to
reagentfillings.dmi

Jalleo commented:
@iamgoofball just to know what you think about this.

Iamgoofball commented:
dicks out for harambe

nullbear commented:
y does this take so long 2 merge q_q

optimumtact commented:
If this breaks chemistry horribly I'm going to delete you

optimumtact commented:
So this broke master and was reverted
code/modules/reagents/chemistry/recipes/slime_extracts.dm:84:error: result: undefined var

code/modules/reagents/chemistry/recipes/slime_extracts.dm:86:error: result_amount: undefined var

PKPenguin321 commented:
ack

nullbear commented:
bleh why are slimes silly.

nullbear commented:
Ah, it's probably a merge conflict or something that isn't showing up, because the code runs fine on my end.
